### ISSUE TITLE
libssl.so.3: cannot open shared object file

### ISSUE THREAD

**Issue by ktmeaton** [phac-nml/rebar#6]
Now that we're in the land of compiled languages (Rust) it's type to troubleshoot dynamic library linking on other systems.

```text
error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
```

**Comment by ktmeaton** [comment]
`rebar: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), statically linked, no section header`

From [StackOverflow](https://stackoverflow.com/a/31778003) by Steve Klabnik: _"Rust statically links everything but glibc (and libgcc, iirc) by default. If you want to get a 100% statically linked binary, you can use MUSL with 1.1. https://github.com/rust-lang/rust/pull/24777 is the initial support, we hope to make it much easier to use in the future._

**Comment by ktmeaton** [comment]
Also [from Kornel](https://users.rust-lang.org/t/unable-to-run-compiled-program/88441/2): _openssl-sys and other openssl-dependent crates have vendored-openssl Cargo feature that you can enable and this usually helps with this particular dependency._

**Comment by ktmeaton** [comment]
The openssl errors can be fixed by using the `vendored` featured of the openssl crate. This will compile OpenSSL from source, and statically include the libraries into the binary

```yaml
openssl = { version = "0.10.59", features = ["vendored"] }
```

**Comment by ktmeaton** [comment]
However! Now there are errors linking to GLIBC on older systems (ex. CentOS 7).

- CentOS 7 uses GLIBC 2.17 ([link](https://www.liquidweb.com/kb/how-to-check-the-glibc-gnu-libc-version-on-centos-6-and-centos-7/)).
- GitHub Actions `ubuntu-latest` uses GLIBC 2.XX?
- My dev environment uses GLIBC 2.35 (`ldd --version ldd`)
```bash
objdump -T target/x86_64-unknown-linux-gnu/release/rebar | tail -n+5 | cut -f 2 | cut -d " " -f 2- | sort
```
- I could try and build for MUSL, which will remove the dependencies on GLIBC. But who knows how big the final binary will be, and if I'll be able to get it to compile at all..
- I might instead try the [cross](https://crates.io/crates/cross) crate, which uses Docker Containers to help with cross-compilation.

**Comment by ktmeaton** [comment]
Resolved with 5dd2ec782ad7b2df4e6d2b582cbaf3e186860c0b by using the cross crate and comping for MUSL!
